### PR TITLE
unit-tests: Unit tests for functions in erasure-readfile.go

### PR DESCRIPTION
**Pull request by krishnasrinivas**
@harshavardhana there is no easy way to unit test erasureReadFile and parallelRead.

The only way is to do xl.PutObject and use the resulting environment to test erasureReadFile and parallelRead, no other way right?

### CODE REVIEW

File: erasure-readfile_test.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	}
+	disks[0] = nil
+	// With one data disk down, it should return dataBlocks-1 data disks and 2 parity disks.
+	readDisks, _, err = getReadDisks(disks, 0, xl.dataBlocks)
+	if err != nil {
+		t.Errorf("getReadDisks returned error: %s", err)
+		return
+	}
+
+	for index := 0; index < len(readDisks); index++ {
```
Comment (harshavardhana) [inline erasure-readfile_test.go]: This whole block needs comments..

File: erasure-readfile_test.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import "testing"
+
+// Test for getReadDisks.
```
Comment (harshavardhana) [inline erasure-readfile_test.go]: Comment should be about 'non-obvious' - should be about the gist of the function body.

File: erasure-readfile_test.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+			} else {
+				if readDisks[index] != nil {
+					t.Errorf("readDisks[%d] should be nil", index)
+				}
+			}
+		}
+	}
+}
+
+// Test getOrderedDisks.
```
Comment (harshavardhana) [inline erasure-readfile_test.go]: Same here.

File: erasure-readfile_test.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+func testGetOrderedDisks(t *testing.T, xl xlObjects) {
+	disks := xl.storageDisks
+	blockCheckSums := make([]checkSumInfo, len(disks))
+	distribution := []int{16, 14, 12, 10, 8, 6, 4, 2, 1, 3, 5, 7, 9, 11, 13, 15}
+	orderedDisks, _ := getOrderedDisks(distribution, disks, blockCheckSums)
+	if orderedDisks[0] != disks[8] ||
```
Comment (harshavardhana) [inline erasure-readfile_test.go:118]: Comment here.

File: erasure-readfile_test.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	enBlocks[dataBlocks-1] = nil
+	// When only parity blocks are available.
+	if isSuccessDecodeBlocks(enBlocks, dataBlocks) == true {
+		t.Errorf("isSuccessDecodeBlocks should return false.")
+	}
+}
+
+// Wrapper function for testGetReadDisks, testGetOrderedDisks.
+func TestErasureReadFile(t *testing.T) {
+	objLayer, dirs, err := getXLObjectLayer()
```
Comment (hackintoshrao) [inline erasure-readfile_test.go:198]: can you use `makeTestBackend` function from `test-utils_test.go`, its a wrapper and used around in the tests/benchmarks.

**Comment by hackintoshrao** [comment]
- It'll be good if unit tests follow the style of formulating inputs and expected output in an array of structs, the cases should contain erroneous inputs and the errors should be verified, that'll make the tests easily extensibe, it'll be more encouraging to add more test cases in the future.
- The exhaustive of the test can be verified by deleting any `if err!=nil` conditional check of errors and see whether the test catches them. Basically test should also work as code preservatives, any undesirable code changes should be caught by the tests ( thats the challenge I'm witnessing w.r.t to the current implementation of locks, not able to write test to catch the removal or addition of any lock related changes in the code).

**Comment by harshavardhana** [comment]
> @harshavardhana there is no easy way to unit test erasureReadFile and parallelRead.
> 
> The only way is to do xl.PutObject and use the resulting environment to test erasureReadFile and parallelRead, no other way right?

This is acceptable, once we have an object through PutObject.
-  We can corrupt them by appending to the file directly using posix.  
-  We can remove the disks entirely.
-  We can remove all but one data disks to hit the parallel read special case. 

All these cases can be simulated and we should do it otherwise we won't be able to catch issues.

**Pull request merged**